### ISSUE TITLE
Add a way to edit ARB files (terms export, files synchronization)

### ISSUE THREAD

**Issue by orestesgaolin** [Dart-Code/Dart-Code#3112]
**Is your feature request related to a problem? Please describe.**

With new l10n approach the only thing to worry about is to update .arb files in `l10n` directory. It's not ideal to edit the files manually and keep them in sync between various languages.

**Describe the solution you'd like**

As a user I would love to be able to:

- have a quick action to export selected string to template arb file
![image](https://user-images.githubusercontent.com/16854239/106457669-99e5b080-648f-11eb-9328-f32cc2bb730f.png)
- automatically sync the terms between all present files (e.g. after adding to template, there should be new entries for all the other languages with empty values)
- add new arb file for desired language (e.g. by providing language code in prompt) that would automatically have all the entries present and empty
- notify me about missing translation entries when opening arb files (e.g. as analyzer info messages)

**Describe alternatives you've considered**

Similar feature is possible with Flutter Intl VS Code extension but it's tightly coupled with their localization system (e.g. generating custom localization delegate, sync with remote server).

**Comment by DanTup** [comment]
It looks like there's a tool that already does something like this:

https://pub.dev/packages/intl_translation#extracting-and-using-translated-messages

> pub run intl_translation:extract_to_arb --output-dir=target/directory my_program.dart more_of_my_program.dart

Does that do what you need? If so, perhaps it could be made easier to run from VS Code (at least if there's a convention about paths etc.)?

**Comment by orestesgaolin** [comment]
> Does that do what you need? If so, perhaps it could be made easier to run from VS Code (at least if there's a convention about paths etc.)?

No, it's something different. This assumes that there's a class containing `intl`-based entries like:

```dart
  String get appBarTitle {
    return Intl.message(
      'Sign Up',
      name: 'appBarTitle',
      desc: 'App bar title shown on Sign Up page',
    );
  }
```

However, with new [l10n approach](https://flutter.dev/docs/development/accessibility-and-localization/internationalization) there's no longer need to generate these manually, as it's done by the `flutter_localizations` in the background (these localization files are generated in `.dart-tool` directory after `pub get`, and the source files are based on the `l10n.yaml` config file and `arb` files in specified directory). 

What I would like to have can be shown in this video:

https://user-images.githubusercontent.com/16854239/106932265-d4fa1500-6717-11eb-8fe8-b88511fa36b7.mp4

This action would result in new entry added in arb file, triggering `flutter_localizations` to regenerate localization classes, and string being replaced by this new key.

I started to play around on my own with sample VS Code extension to do it, but it would be awesome to have official support to even further simplify the localization process.

**Comment by DanTup** [comment]
@HansMuller I'm told you might be familiar with the localisation tools here so you might have some ideas about the current best practices. It sounds like the `intl_translation:extract_to_arb` tool works only with `Intl.message()` calls, though it may be more common to now use `AppLocalizations.of(context).helloWorld` instead?

Is it possible/likely that `intl_translation:extract_to_arb` would be extended in some way to support extraction using that format? (it seems like it might be difficult to extract them without all the info that was in the previous `Intl.message` calls?).

Are the file formats/locations static enough to have assists as shown in the screenshot at the top to add new strings to them? If the analysis server had some knowledge of these files, perhaps it could also have additional functionality such as go-to-definition and find-references working with the source files and not just the generated Dart; although there may be value in adding something simpler directly into VS Code in the short-term for the simple extraction case.

**Comment by DanTup** [comment]
@HansMuller any thoughts on the above?

**Comment by HansMuller** [comment]
@DanTup - sorry to have gone dark here, I'd overlooked [your comment on 2/11](https://github.com/Dart-Code/Dart-Code/issues/3112#issuecomment-777630350). Our current recommendation for localizing Flutter apps is described here: https://flutter.dev/docs/development/accessibility-and-localization/internationalization. It still begins with .arb files but is now based on an automatically generated localizations class and is not tied to the Dart intl package's tools.

**Comment by DanTup** [comment]
Sorry for the long delay with this. I discussed this with someone on the Dart team and agreed the place to do something like this would be in the analysis server, by having it know about these arb files and be able to produce edits to add strings to them.

I've raised an issue at https://github.com/dart-lang/sdk/issues/46722 with some details from this issue (feel free to add anything I missed) and will close this one in favour of that. It would be quite difficult to do directly here in the extension as it doesn't have a good AST of the code (plus any solution here would only work for VS Code and not IntelliJ/Android Studio or any of the other LSP clients). Hope this makes sense!
